internal/codeintel/db/reference_pager.go

```
@@ -1,39 +1,46 @@
 package db
 
 import (
-	"database/sql"
+	"context"
 
 	"github.com/sourcegraph/sourcegraph/internal/codeintel/bundles/types"
 )
 
-// ReferencePager holds state for a reference result in a SQL transaction. Each page
-// requested should have a consistent view into the database.
+// ReferencePager holds state for a reference result within a SQL transaction so that
+// each page requested has a consistent view into the database.
 type ReferencePager interface {
-	TxCloser
-
 	// PageFromOffset returns the page of package references that starts at the given offset.
-	PageFromOffset(offset int) ([]types.PackageReference, error)
+	PageFromOffset(ctx context.Context, offset int) ([]types.PackageReference, error)
+
+	// Done closes the underlying transaction. If the reference pager was called on a
+	// Database instance that was already in a transaction, this method does nothing.
+	Done(err error) error
+}
```

suggestion:
instead of returning a `Done` function, can we use closures to guarantee the transaction gets closed? or would that make using this really awkward?

i.e. `PageFromOffset(func (PageFromOffsetFn))` and then delete the `Done` field.

i'm used to environments where not closing txs is a Big Deal but i know there exist more robust systems than that, so just lmk if i'm being too nitpicky about this

---

It's actually more idiomatic to return done as we can defer it right after successful creation.

---

Also if you look at the usage, we call PageFromOffset multiple times in a loop and Done only when we want to close the underlying txn.

---

right, the usage would be
```
pager.PageFromOffset(func (pageFromOffsetFn) {
    for whatever {
      refs, err := pageFromOffsetFn(ctx, offset)
    }
}
```
happy with either, but i will probably forget to defer at some point so you better find it in review lol

---

Oh I see. I think we try not to design APIs with closure state like that as you need to close over more than you would usually in languages with better support for generics/var args. For example we can't return anything (or need to return a single error type, or interface{} and error) and if the loop is meant to calculate something to return you need to modify "global"ish state.

---

not sure i understand the second point, couldn't you
```golang
var return []types.PackageReference
pager.PageFromOffset(func (pageFromOffsetFn) {
  for whatever {
    refs, err := pageFromOffsetFn(ctx, offset)
    return = append(return, ...refs)
  }
})
```
Go idioms generally frazzle my FP brain so i'm probably swimming upstream here, go with whatever makes more sense to you here

---

Yes, but that's precisely the gross "global"ish state.

---

ah, i wouldn't consider that global-ish, but i understand what you're saying

---

That's why I said _isssshhhh_ 😛

---

it's totally local-ish tho! "ish" is a very technical term and i won't stand for its misuse